src/haz3lcore/statics/TypBase.re

```diff
@@ -175,40 +224,22 @@ module rec Typ: {
     | (String, _) => false
     | (Unknown(_), Unknown(_)) => true
     | (Unknown(_), _) => false
-    | (Arrow(t1, t2), Arrow(t1', t2')) => eq(t1, t1') && eq(t2, t2')
+    | (Arrow(t1, t2), Arrow(t1', t2')) =>
+      eq_internal(n, t1, t1') && eq_internal(n, t2, t2')
     | (Arrow(_), _) => false
-    | (Prod(tys1), Prod(tys2)) => List.equal(eq, tys1, tys2)
+    | (Prod(tys1), Prod(tys2)) => List.equal(eq_internal(n), tys1, tys2)
     | (Prod(_), _) => false
-    | (List(t1), List(t2)) => eq(t1, t2)
+    | (List(t1), List(t2)) => eq_internal(n, t1, t2)
     | (List(_), _) => false
     | (Sum(sm1), Sum(sm2)) =>
-      ConstructorMap.equal(Option.equal(eq), sm1, sm2)
+      ConstructorMap.equal(Option.equal(eq_internal(n)), sm1, sm2)
     | (Sum(_), _) => false
     | (Var(n1), Var(n2)) => n1 == n2
```

Comment:
this does not normalize singletons, so that should be noted in a comment -- this is only type equivalence on already normalized types